### ISSUE TITLE
Expand component configurable logging system to C++

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#98872]
The component configurable logging system introduced for #94788 needs to be expanded to control logs generated from C++

Some of the requirements for this are discussed here:
* https://github.com/pytorch/pytorch/issues/94788#issuecomment-1502519872
* https://github.com/pytorch/pytorch/issues/94788#issuecomment-1503989249

**Comment by kurtamohler** [comment]
I think it would make sense to make a `PyLogHandler` similar to `PyWarningHandler` in `torch/csrc/Exceptions.h`. `HANDLE_TH_ERRORS` would just instantiate the handler, and when the handler is destructed in `END_HANDLE_TH_ERRORS`, any logs that were added to a temporary log buffer during the operation will get sent to Python.

Sending a log to Python will require writing CPython code that imports the `logging` module and calls the appropriate `logging.warning`, `logging.debug`, etc function.

I will start working on a draft implementation

**Comment by lezcano** [comment]
cc @ezyang

**Comment by ezyang** [comment]
This all seems reasonable. I'd add that I think it's important to be able to specify source files specifically for logging, which means we'll need to support something like TORCH_LOGS=aten/src/ATen/Foo.cpp which is different from what we do today in Python.

**Comment by mlazos** [comment]
Agreed with ed's file support comment, otherwise this sounds very good, I will take a look at the PR tomorrow. Thanks for doing this!
